File: languages/python/calculators/Binomial Expansion.py
# (a + b) ^ n
superscript = ['⁰', '¹', '²', '³', '⁴', '⁵', '⁶', '⁷', '⁸', '⁹']

def factorial(n):
    ans = 1
    for i in range(2, n + 1):
        ans *= i
    return ans

def nCr(n, r):
    return factorial(n) / (factorial(r) * factorial(n - r))

while True:
    try:
        n = int(input('Please enter a power to put (a + b) to! '))
        break
    except ValueError:
        print('The power must be a valid integer!')
        continue

terms = []

for r in range(n + 1):
    a = n - r
    b = r
    c = int(nCr(n, r))
    term = ''

    if c != 1:
        term += str(c)

    if a != 0:
        term += 'a'
        if a != 1:
            power = ''
            for digit in str(a):
                power += superscript[int(digit)]
            term += power

    if b != 0:
        term += 'b'
        if b != 1:
            power = ''
            for digit in str(b):
                power += superscript[int(digit)]
            term += power

    terms.append(term)

print(' + '.join(terms))

File: c++/code/binomialExpansion.cc
#include <iostream>

std::string superscript[10] = {"⁰", "¹", "²", "³", "⁴", "⁵", "⁶", "⁷", "⁸", "⁹"};

int factorial(int number) {
	int out = 1;

	for (int i = 2; i <= number; i++) {
		out *= i;
	}

	return out;
}

int nCr(int n, int r) {
	return factorial(n) / (factorial(r) * factorial(n - r));
}

std::string formatNomial(int value, std::string nomial) {
	std::string term = nomial;

	if (value != 1) {
		std::string power = "";

		for (char & digit : std::to_string(value)) {
			power += superscript[digit - '0'];
		}

		term += power;
	}

	return term;
}

std::string binomialExpansion(int power) {
	std::string output = "";

	for (int r = 0; r <= power; r++) {
		int a = power - r;
		int c = nCr(power, r);
		std::string term = "";

		if (c != 1) {
			term += std::to_string(c);
		}

		if (a != 0) {
			term += formatNomial(a, "a");
		}

		if (r != 0) {
			term += formatNomial(r, "b");
		}

		if (r != 0) {
			output += " + " + term;
		} else {
			output += term;
		}
	}

	return output;
}

int main() {
	std::cout << "(a + b)ⁿ\nPlease input a value for n! ";

	int n;
	std::cin >> n;

	std::cout << binomialExpansion(n);
}
